ProducerMasterChangeTest

```
[ExecutorGraph-materialize-id-0-0] ERROR org.apache.zookeeper.server.NIOServerCnxnFactory - Thread Thread[ExecutorGraph-materialize-id-0-0,5,main] died
java.lang.IllegalStateException: Executor com.bwsw.tstreams.common.FirstFailLockableTaskExecutor@24243111[Shutting down, pool size = 1, active threads = 1, queued tasks = 0, completed tasks = 590] is failed. No new tasks are allowed.
	at com.bwsw.tstreams.coordination.producer.ExecutorGraph$$anonfun$submitTo$1.apply(ExecutorGraph.scala:38)
	at com.bwsw.tstreams.coordination.producer.ExecutorGraph$$anonfun$submitTo$1.apply(ExecutorGraph.scala:38)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at com.bwsw.tstreams.coordination.producer.ExecutorGraph.submitTo(ExecutorGraph.scala:38)
	at com.bwsw.tstreams.coordination.producer.ExecutorGraph.submitToPublish(ExecutorGraph.scala:53)
	at com.bwsw.tstreams.coordination.producer.PeerAgent.submitPipelinedTaskToPublishExecutors(PeerAgent.scala:414)
	at com.bwsw.tstreams.agents.producer.Producer.openTransactionLocal(Producer.scala:265)
	at com.bwsw.tstreams.coordination.messages.master.NewTransactionRequest$$anonfun$handleP2PRequest$1.apply$mcV$sp(Messages.scala:86)
	at com.bwsw.tstreams.coordination.producer.ExecutorGraph$$anonfun$submitTo$2$$anon$1.run(ExecutorGraph.scala:41)
	at com.bwsw.tstreams.common.FirstFailLockableTaskExecutor$RunnableWithoutLock.run(FirstFailLockableTaskExecutor.scala:82)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
java.lang.AssertionError: assertion failed
	at scala.Predef$.assert(Predef.scala:156)
	at com.bwsw.tstreams.agents.producer.Producer.materialize(Producer.scala:348)
	at com.bwsw.tstreams.coordination.messages.master.MaterializeRequest.handleP2PRequest(Messages.scala:301)
	at com.bwsw.tstreams.coordination.messages.master.IMessage$class.run(Messages.scala:51)
	at com.bwsw.tstreams.coordination.messages.master.MaterializeRequest.run(Messages.scala:297)
	at com.bwsw.tstreams.coordination.producer.PeerAgent$$anonfun$2.apply$mcV$sp(PeerAgent.scala:390)
	at com.bwsw.tstreams.coordination.producer.ExecutorGraph$$anonfun$submitTo$2$$anon$1.run(ExecutorGraph.scala:41)
	at com.bwsw.tstreams.common.FirstFailLockableTaskExecutor$RunnableWithoutLock.run(FirstFailLockableTaskExecutor.scala:82)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

```

```
[ExecutorGraph-materialize-id-0-0] WARN com.bwsw.tstreams.common.FirstFailLockableTaskExecutor - Task <ExecutorGraphTask> :: <MaterializeTask> has delayed Full in executor ExecutorGraph-materialize-id-0 for 3034 msecs. Threshold is: 120 msecs.
[ExecutorGraph-materialize-id-0-0] WARN com.bwsw.tstreams.common.FirstFailLockableTaskExecutor - Task <ExecutorGraphTask> :: <MaterializeTask> has run in executor ExecutorGraph-materialize-id-0 for 3034 msecs. Threshold is: 20 msecs.
[ExecutorGraph-materialize-id-0-0] ERROR com.bwsw.tstreams.common.FirstFailLockableTaskExecutor - scala.Predef$.assert(Predef.scala:156)
[ExecutorGraph-materialize-id-0-0] ERROR com.bwsw.tstreams.common.FirstFailLockableTaskExecutor - com.bwsw.tstreams.agents.producer.Producer.materialize(Producer.scala:348)
[ExecutorGraph-materialize-id-0-0] ERROR com.bwsw.tstreams.common.FirstFailLockableTaskExecutor - com.bwsw.tstreams.coordination.messages.master.MaterializeRequest.handleP2PRequest(Messages.scala:301)
[ExecutorGraph-materialize-id-0-0] ERROR com.bwsw.tstreams.common.FirstFailLockableTaskExecutor - com.bwsw.tstreams.coordination.messages.master.IMessage$class.run(Messages.scala:51)
[ExecutorGraph-materialize-id-0-0] ERROR com.bwsw.tstreams.common.FirstFailLockableTaskExecutor - com.bwsw.tstreams.coordination.messages.master.MaterializeRequest.run(Messages.scala:297)
[ExecutorGraph-materialize-id-0-0] ERROR com.bwsw.tstreams.common.FirstFailLockableTaskExecutor - com.bwsw.tstreams.coordination.producer.PeerAgent$$anonfun$2.apply$mcV$sp(PeerAgent.scala:390)
[ExecutorGraph-materialize-id-0-0] ERROR com.bwsw.tstreams.common.FirstFailLockableTaskExecutor - com.bwsw.tstreams.coordination.producer.ExecutorGraph$$anonfun$submitTo$2$$anon$1.run(ExecutorGraph.scala:41)
[ExecutorGraph-materialize-id-0-0] ERROR com.bwsw.tstreams.common.FirstFailLockableTaskExecutor - com.bwsw.tstreams.common.FirstFailLockableTaskExecutor$RunnableWithoutLock.run(FirstFailLockableTaskExecutor.scala:82)
[ExecutorGraph-materialize-id-0-0] ERROR com.bwsw.tstreams.common.FirstFailLockableTaskExecutor - java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
[ExecutorGraph-materialize-id-0-0] ERROR com.bwsw.tstreams.common.FirstFailLockableTaskExecutor - java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
[ExecutorGraph-materialize-id-0-0] ERROR com.bwsw.tstreams.common.FirstFailLockableTaskExecutor - java.lang.Thread.run(Thread.java:745)
[ExecutorGraph-materialize-id-0-0] ERROR org.apache.zookeeper.server.NIOServerCnxnFactory - Thread Thread[ExecutorGraph-materialize-id-0-0,5,main] died
java.lang.AssertionError: assertion failed
	at scala.Predef$.assert(Predef.scala:156)
	at com.bwsw.tstreams.agents.producer.Producer.materialize(Producer.scala:348)
	at com.bwsw.tstreams.coordination.messages.master.MaterializeRequest.handleP2PRequest(Messages.scala:301)
	at com.bwsw.tstreams.coordination.messages.master.IMessage$class.run(Messages.scala:51)
	at com.bwsw.tstreams.coordination.messages.master.MaterializeRequest.run(Messages.scala:297)
	at com.bwsw.tstreams.coordination.producer.PeerAgent$$anonfun$2.apply$mcV$sp(PeerAgent.scala:390)
	at com.bwsw.tstreams.coordination.producer.ExecutorGraph$$anonfun$submitTo$2$$anon$1.run(ExecutorGraph.scala:41)
	at com.bwsw.tstreams.common.FirstFailLockableTaskExecutor$RunnableWithoutLock.run(FirstFailLockableTaskExecutor.scala:82)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
java.lang.IllegalStateException: Executor com.bwsw.tstreams.common.FirstFailLockableTaskExecutor@3f110ead[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 622] is failed. No new tasks are allowed.
	at com.bwsw.tstreams.coordination.producer.ExecutorGraph$$anonfun$submitTo$1.apply(ExecutorGraph.scala:38)
	at com.bwsw.tstreams.coordination.producer.ExecutorGraph$$anonfun$submitTo$1.apply(ExecutorGraph.scala:38)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at com.bwsw.tstreams.coordination.producer.ExecutorGraph.submitTo(ExecutorGraph.scala:38)
	at com.bwsw.tstreams.coordination.producer.ExecutorGraph.submitToPublish(ExecutorGraph.scala:53)
	at com.bwsw.tstreams.coordination.producer.PeerAgent.submitPipelinedTaskToPublishExecutors(PeerAgent.scala:414)
	at com.bwsw.tstreams.agents.producer.Producer.openTransactionLocal(Producer.scala:265)
	at com.bwsw.tstreams.coordination.messages.master.NewTransactionRequest$$anonfun$handleP2PRequest$1.apply$mcV$sp(Messages.scala:86)
	at com.bwsw.tstreams.coordination.producer.ExecutorGraph$$anonfun$submitTo$2$$anon$1.run(ExecutorGraph.scala:41)
	at com.bwsw.tstreams.common.FirstFailLockableTaskExecutor$RunnableWithoutLock.run(FirstFailLockableTaskExecutor.scala:82)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
[ExecutorGraph-cassandra-id-0-0] ERROR com.bwsw.tstreams.common.FirstFailLockableTaskExecutor - com.bwsw.tstreams.coordination.producer.ExecutorGraph$$anonfun$submitTo$1.apply(ExecutorGraph.scala:38)
[ExecutorGraph-cassandra-id-0-0] ERROR com.bwsw.tstreams.common.FirstFailLockableTaskExecutor - com.bwsw.tstreams.coordination.producer.ExecutorGraph$$anonfun$submitTo$1.apply(ExecutorGraph.scala:38)
[ExecutorGraph-cassandra-id-0-0] ERROR com.bwsw.tstreams.common.FirstFailLockableTaskExecutor - scala.collection.immutable.List.foreach(List.scala:381)
[ExecutorGraph-cassandra-id-0-0] ERROR com.bwsw.tstreams.common.FirstFailLockableTaskExecutor - com.bwsw.tstreams.coordination.producer.ExecutorGraph.submitTo(ExecutorGraph.scala:38)
[ExecutorGraph-cassandra-id-0-0] ERROR com.bwsw.tstreams.common.FirstFailLockableTaskExecutor - com.bwsw.tstreams.coordination.producer.ExecutorGraph.submitToPublish(ExecutorGraph.scala:53)
[ExecutorGraph-cassandra-id-0-0] ERROR com.bwsw.tstreams.common.FirstFailLockableTaskExecutor - com.bwsw.tstreams.coordination.producer.PeerAgent.submitPipelinedTaskToPublishExecutors(PeerAgent.scala:414)
[ExecutorGraph-cassandra-id-0-0] ERROR com.bwsw.tstreams.common.FirstFailLockableTaskExecutor - com.bwsw.tstreams.agents.producer.Producer.openTransactionLocal(Producer.scala:265)
[ExecutorGraph-cassandra-id-0-0] ERROR com.bwsw.tstreams.common.FirstFailLockableTaskExecutor - com.bwsw.tstreams.coordination.messages.master.NewTransactionRequest$$anonfun$handleP2PRequest$1.apply$mcV$sp(Messages.scala:86)
[ExecutorGraph-cassandra-id-0-0] ERROR com.bwsw.tstreams.common.FirstFailLockableTaskExecutor - com.bwsw.tstreams.coordination.producer.ExecutorGraph$$anonfun$submitTo$2$$anon$1.run(ExecutorGraph.scala:41)
[ExecutorGraph-cassandra-id-0-0] ERROR com.bwsw.tstreams.common.FirstFailLockableTaskExecutor - com.bwsw.tstreams.common.FirstFailLockableTaskExecutor$RunnableWithoutLock.run(FirstFailLockableTaskExecutor.scala:82)
[ExecutorGraph-cassandra-id-0-0] ERROR com.bwsw.tstreams.common.FirstFailLockableTaskExecutor - java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
[ExecutorGraph-cassandra-id-0-0] ERROR com.bwsw.tstreams.common.FirstFailLockableTaskExecutor - java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
[ExecutorGraph-cassandra-id-0-0] ERROR com.bwsw.tstreams.common.FirstFailLockableTaskExecutor - java.lang.Thread.run(Thread.java:745)
[ExecutorGraph-cassandra-id-0-0] ERROR org.apache.zookeeper.server.NIOServerCnxnFactory - Thread Thread[ExecutorGraph-cassandra-id-0-0,5,main] died
java.lang.IllegalStateException: Executor com.bwsw.tstreams.common.FirstFailLockableTaskExecutor@3f110ead[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 622] is failed. No new tasks are allowed.
	at com.bwsw.tstreams.coordination.producer.ExecutorGraph$$anonfun$submitTo$1.apply(ExecutorGraph.scala:38)
	at com.bwsw.tstreams.coordination.producer.ExecutorGraph$$anonfun$submitTo$1.apply(ExecutorGraph.scala:38)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at com.bwsw.tstreams.coordination.producer.ExecutorGraph.submitTo(ExecutorGraph.scala:38)
	at com.bwsw.tstreams.coordination.producer.ExecutorGraph.submitToPublish(ExecutorGraph.scala:53)
	at com.bwsw.tstreams.coordination.producer.PeerAgent.submitPipelinedTaskToPublishExecutors(PeerAgent.scala:414)
	at com.bwsw.tstreams.agents.producer.Producer.openTransactionLocal(Producer.scala:265)
	at com.bwsw.tstreams.coordination.messages.master.NewTransactionRequest$$anonfun$handleP2PRequest$1.apply$mcV$sp(Messages.scala:86)
	at com.bwsw.tstreams.coordination.producer.ExecutorGraph$$anonfun$submitTo$2$$anon$1.run(ExecutorGraph.scala:41)
	at com.bwsw.tstreams.common.FirstFailLockableTaskExecutor$RunnableWithoutLock.run(FirstFailLockableTaskExecutor.scala:82)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
сен 16, 2016 2:58:23 PM com.twitter.common.zookeeper.DistributedLockImpl unlock
INFO: Cleaning up this locks ephemeral node.
сен 16, 2016 2:58:23 PM com.twitter.common.zookeeper.DistributedLockImpl cleanup

```

--- a/src/main/scala-2.11/com/bwsw/tstreams/agents/producer/IProducerTransaction.scala
+++ b/src/main/scala-2.11/com/bwsw/tstreams/agents/producer/IProducerTransaction.scala
@@ -25,4 +25,6 @@ trait IProducerTransaction[T] {
   def getTransactionID(): Long
 
   def makeMaterialized(): Unit
+
+  def markAsClosed(): Unit
 }

--- a/src/main/scala-2.11/com/bwsw/tstreams/agents/producer/Producer.scala
+++ b/src/main/scala-2.11/com/bwsw/tstreams/agents/producer/Producer.scala
@@ -90,7 +90,7 @@ class Producer[T](var name: String,
       pcs.threadPoolAmount
   }
 
-  stream.dataStorage.bind() //TODO: fix, probably deprecated
+  stream.dataStorage.bind()
 
   logger.info(s"Start new Basic producer with name : $name, streamName : ${stream.getName}, streamPartitions : ${stream.getPartitions}")
 
@@ -169,12 +169,7 @@ class Producer[T](var name: String,
     openTransactions.forallKeysDo((part: Int, transaction: IProducerTransaction[T]) => transaction.updateTransactionKeepAliveState())
   }
 
-  /**
-    * @param policy    Policy for previous transaction on concrete partition
-    * @param partition Next partition to use for transaction (default -1 which mean that write policy will be used)
-    * @return BasicProducerTransaction instance
-    */
-  def newTransaction(policy: ProducerPolicy, partition: Int = -1): ProducerTransaction[T] = {
+  private def newTransactionUnsafe(policy: ProducerPolicy, partition: Int = -1): ProducerTransaction[T] = {
     if (isStop)
       throw new IllegalStateException(s"Producer ${this.name} is already stopped. Unable to get new transaction.")
 
@@ -208,6 +203,29 @@ class Producer[T](var name: String,
       })
     transaction
   }
+
+  /**
+    * @param policy    Policy for previous transaction on concrete partition
+    * @param partition Next partition to use for transaction (default -1 which mean that write policy will be used)
+    * @return BasicProducerTransaction instance
+    */
+  def newTransaction(policy: ProducerPolicy, partition: Int = -1, retry: Int = 1): ProducerTransaction[T] = {
+    if (isStop)
+      throw new IllegalStateException(s"Producer ${this.name} is already stopped. Unable to get new transaction.")
+
+    if(retry < 0)
+      throw new IllegalStateException("Failed to get a new transaction.")
+
+    try {
+      val transaction = newTransactionUnsafe(policy, partition)
+      transaction.awaitMaterialized()
+      transaction
+    } catch {
+      case e: MaterializationException =>
+        newTransaction(policy, retry - 1)
+    }
+  }
+
 
   /**
     * Return transaction for specific partition if there is opened one.
@@ -347,16 +365,28 @@ class Producer[T](var name: String,
     *
     * @param msg
     */
-  def materialize(msg: TransactionStateMessage) = {
+  def materialize(msg: TransactionStateMessage):Unit = {
+
     if (logger.isDebugEnabled)
       logger.debug(s"Start handling MaterializeRequest at partition: ${msg.partition}")
+
     materializationGovernor.awaitUnprotected(msg.partition)
     val opt = getOpenedTransactionForPartition(msg.partition)
-    assert(opt.isDefined)
+
+    if(opt.isEmpty) {
+      logger.warn(s"There is no opened transaction for ${msg.partition}.")
+      return
+    }
+
     if (logger.isDebugEnabled)
       logger.debug(s"In Map Transaction: ${opt.get.getTransactionID.toString}\nIn Request Transaction: ${msg.transactionID}")
-    assert(opt.get.getTransactionID == msg.transactionID)
-    assert(msg.status == TransactionStatus.materialize)
+
+    if(!(opt.get.getTransactionID == msg.transactionID && msg.status == TransactionStatus.materialize)) {
+      logger.warn(s"Materialization is requested for transaction ${msg.transactionID} but expected transaction is ${opt.get.getTransactionID}.")
+      opt.get.markAsClosed()
+      return
+    }
+
     opt.get.makeMaterialized()
     if (logger.isDebugEnabled)
       logger.debug("End handling MaterializeRequest")

--- a/src/main/scala-2.11/com/bwsw/tstreams/agents/producer/ProducerTransaction.scala
+++ b/src/main/scala-2.11/com/bwsw/tstreams/agents/producer/ProducerTransaction.scala
@@ -1,6 +1,7 @@
 package com.bwsw.tstreams.agents.producer
 
 import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.AtomicBoolean
 import java.util.concurrent.locks.ReentrantLock
 
 import com.bwsw.tstreams.agents.group.ProducerCheckpointInfo

--- a/src/main/scala-2.11/com/bwsw/tstreams/coordination/producer/AgentsStateDBService.scala
+++ b/src/main/scala-2.11/com/bwsw/tstreams/coordination/producer/AgentsStateDBService.scala
@@ -92,7 +92,7 @@ class AgentsStateDBService(dlm: ZookeeperDLMService,
       partitions foreach { p =>
         val penalty = if (isLowPriorityToBeMaster) PeerAgent.LOW_PRIORITY_PENALTY else 0
         val conf = AgentConfiguration(inetAddress, weight = 0, penalty, uniqueAgentId)
-        println(s"ConfInit: $conf")
+        //println(s"ConfInit: $conf")
         dlm.create[AgentConfiguration](getMyPath(p), conf, CreateMode.EPHEMERAL)
       }
       if (PeerAgent.logger.isDebugEnabled) {

--- a/src/test/scala-2.11/agents/producer/OpenTransactionsKeeperTests.scala
+++ b/src/test/scala-2.11/agents/producer/OpenTransactionsKeeperTests.scala
@@ -41,25 +41,27 @@ class OpenTransactionsKeeperTests extends FlatSpec with Matchers {
     override def getTransactionID(): Long = LocalGeneratorCreator.getTransaction()
 
     override def makeMaterialized(): Unit = {}
+
+    override def markAsClosed(): Unit = {}
   }
 
   it should "allow add and get transactions to it" in {
-    val otk = new OpenTransactionsKeeper[String]()
-    otk.put(0, new TransactionStub)
-    otk.getTransactionOption(0).isDefined shouldBe true
-    otk.getTransactionOption(1).isDefined shouldBe false
+    val keeper = new OpenTransactionsKeeper[String]()
+    keeper.put(0, new TransactionStub)
+    keeper.getTransactionOption(0).isDefined shouldBe true
+    keeper.getTransactionOption(1).isDefined shouldBe false
   }
 
   it should "handle all variants of awaitMaterialized" in {
-    val otk = new OpenTransactionsKeeper[String]()
+    val keeper = new OpenTransactionsKeeper[String]()
     val t = new TransactionStub
-    otk.put(0, t)
-    otk.awaitOpenTransactionMaterialized(0, NewTransactionProducerPolicy.CheckpointIfOpened)()
+    keeper.put(0, t)
+    keeper.awaitOpenTransactionMaterialized(0, NewTransactionProducerPolicy.CheckpointIfOpened)()
     t.lastMethod shouldBe "checkpoint"
-    otk.awaitOpenTransactionMaterialized(0, NewTransactionProducerPolicy.CancelIfOpened)()
+    keeper.awaitOpenTransactionMaterialized(0, NewTransactionProducerPolicy.CancelIfOpened)()
     t.lastMethod shouldBe "cancel"
     (try {
-      otk.awaitOpenTransactionMaterialized(0, NewTransactionProducerPolicy.ErrorIfOpened)()
+      keeper.awaitOpenTransactionMaterialized(0, NewTransactionProducerPolicy.ErrorIfOpened)()
       false
     } catch {
       case e: IllegalStateException =>
@@ -68,11 +70,11 @@ class OpenTransactionsKeeperTests extends FlatSpec with Matchers {
   }
 
   it should "handle for all keys do properly" in {
-    val otk = new OpenTransactionsKeeper[String]()
-    otk.put(0, new TransactionStub)
-    otk.put(1, new TransactionStub)
-    otk.put(2, new TransactionStub)
-    otk.forallKeysDo[Unit]((p: Int, t: IProducerTransaction[String]) => t.updateTransactionKeepAliveState())
+    val keeper = new OpenTransactionsKeeper[String]()
+    keeper.put(0, new TransactionStub)
+    keeper.put(1, new TransactionStub)
+    keeper.put(2, new TransactionStub)
+    keeper.forallKeysDo[Unit]((p: Int, t: IProducerTransaction[String]) => t.updateTransactionKeepAliveState())
     ctr shouldBe 3
   }
 

